pkgs/applications/window-managers/sway/wrapper.nix

```diff
@@ -1,53 +1,48 @@
-{ lib
-, stdenv
-, sway-unwrapped
-, swaybg
-, dbus
-, makeWrapper
-, wrapGAppsHook
-, symlinkJoin
-, writeShellScriptBin
-, withBaseWrapper ? true
-, withGTKWrapper ? false
-, extraSessionCommands ? ""
+{ lib, stdenv
+, sway-unwrapped, swaybg
+, makeWrapper, symlinkJoin, writeShellScriptBin
+, withBaseWrapper ? true, extraSessionCommands ? "", dbus
+, withGtkWrapper ? false, wrapGAppsHook
 }:
 
+assert extraSessionCommands != "" -> withBaseWrapper;
+
 with lib;
 
 let
-  swayPackage = sway-unwrapped;
-
   baseWrapper = writeShellScriptBin "sway" ''
-    set -o errexit
-
-    if [ ! "$_SWAY_WRAPPER_ALREADY_EXECUTED" ]; then
-      export _SWAY_WRAPPER_ALREADY_EXECUTED=1
-      ${extraSessionCommands}
-    fi
-
-    if [ "$DBUS_SESSION_BUS_ADDRESS" ]; then
-      export DBUS_SESSION_BUS_ADDRESS
-      exec ${swayPackage}/bin/sway "$@"
-    else
-      exec ${dbus}/bin/dbus-run-session ${swayPackage}/bin/sway "$@"
-    fi
-  '';
-in
-symlinkJoin {
+     set -o errexit
+     if [ ! "$_SWAY_WRAPPER_ALREADY_EXECUTED" ]; then
+       export _SWAY_WRAPPER_ALREADY_EXECUTED=1
+       ${extraSessionCommands}
+     fi
+     if [ "$DBUS_SESSION_BUS_ADDRESS" ]; then
+       export DBUS_SESSION_BUS_ADDRESS
+       exec ${sway-unwrapped}/bin/sway "$@"
+     else
+       exec ${dbus}/bin/dbus-run-session ${sway-unwrapped}/bin/sway "$@"
+     fi
+   '';
+in symlinkJoin {
   name = "sway-${sway-unwrapped.version}";
 
-  paths = (optional withBaseWrapper [baseWrapper])
-    ++ [swayPackage];
+  paths = (optional withBaseWrapper baseWrapper)
+    ++ [ sway-unwrapped ];
 
-  nativeBuildInputs = [makeWrapper]
-    ++ (optional withGTKWrapper [wrapGAppsHook]);
+  nativeBuildInputs = [ makeWrapper ]
+    ++ (optional withGtkWrapper wrapGAppsHook);
 
-  buildInputs = sway-unwrapped.buildInputs;
+  buildInputs = sway-unwrapped.buildInputs; # Required for wrapGAppsHook
+
+  # We want to run wrapProgram manually (with additional parameters)
+  dontWrapGApps = true;
 
   postBuild = ''
-    wrapProgram $out/bin/sway --prefix PATH : "${swaybg}/bin"
-  ''
-  + optionalString withGTKWrapper "wrapGAppsHook";
+    wrapProgram $out/bin/sway \
+      --prefix PATH : "${swaybg}/bin" ${optionalString withGtkWrapper ''\
+        "''${gappsWrapperArgs[@]}"
+      ''}
```

Comment:
This is pretty dirty tbh... :o The result should be fine, but the code is probably difficult to understand... I couldn't find a nicer way to write this (without adding more abstractions).

Reply:
I see, I had to read it twice to understand the intent (but I'm not yet comfortable with all Nix subtleties), maybe we could let a comment here ?